rr_iarrc/src/startlight_watcher/startlight_watcher.cpp

```diff
@@ -0,0 +1,224 @@
+    cv::morphologyEx(greenLightCheck, greenLightCheck, cv::MORPH_OPEN, kernel);
+
+    return greenLightCheck;
+}
+
+//@note circles returned is a vector of vectors containing <x, y, radius>
+std::vector<cv::Vec3f> findCirclesFromContours(cv::Mat &binary) {
+    // Find contours
+    std::vector<std::vector<cv::Point>> contours;
+    std::vector<cv::Vec4i> hierarchy;
+    cv::findContours(binary, contours, hierarchy, CV_RETR_TREE, CV_CHAIN_APPROX_SIMPLE, cv::Point(0, 0));
+
+    // Find circles
+    double circularityThreshold = 0.8; //@note: controls the sensitivity of circle detection
+    std::vector<cv::Vec3f> foundCircles;
+
+    for(int i = 0; i < contours.size(); i++) {
+        double area = cv::contourArea(contours[i]);
+        double arclength = cv::arcLength(contours[i], true);
+        double circularity = 4 * CV_PI * area / (arclength * arclength);
+        if (circularity >= circularityThreshold) {
+            cv::Point2f center;
+            float radius;
+            cv::minEnclosingCircle(contours[i], center, radius);
+            cv::Vec3f circle{center.x, center.y, radius};
+            foundCircles.push_back(circle);
+        }
+    }
+    return foundCircles;
+}
+
+//@note circles returned is a vector of vectors containing <x, y, radius>
+std::vector<cv::Vec3f> findCirclesFromHough(cv::Mat &binary) {
+    /// Apply the Hough Transform to find the circles
+    std::vector<cv::Vec3f> circles;
+    double dp = 1;
+    double minDist = binary.rows/8;
+    double cannyThreshold = 100;
+    double accumThreshold = 12;
+    int minRadius = 0;
+    int maxRadius = binary.rows/3;
+    cv::HoughCircles(binary, circles, CV_HOUGH_GRADIENT, dp, minDist, cannyThreshold, accumThreshold, minRadius, maxRadius);
+    return circles;
+}
+
+//For debugging purposes
+void drawCircles(cv::Mat &debug, std::vector<cv::Vec3f> circles) {
+    // Draw the circles detected
+    for( size_t i = 0; i < circles.size(); i++ ) {
+        cv::Point center(cvRound(circles[i][0]), cvRound(circles[i][1]));
+        int radius = cvRound(circles[i][2]);
+        radius += 0.8 * radius; //add an offset for easier viewing
+        // circle outline
+        cv::circle( debug, center, radius, cv::Scalar(0,255,255), 3, 8, 0 );
+     }
+}
+
+
+
+
+/*
+ * Image callback
+ * The idea: keep checking for green circles, when we find one,
+ * go back [x] frames and see if there was a red one above and nearby.
+ *
+*/
+void img_callback(const sensor_msgs::Image::ConstPtr& msg) {
+
+    if(start_detected.data) {
+        bool_pub.publish(start_detected);
+        return; // do not to all this computing if the signal has already been seen and go broadcast
+    }
+
+    //Convert msg to Mat image
+    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
+    cv::Mat frame = cv_ptr->image;
+
+    //gather a set of frames to run on
+    for(int i = history.size()-1; i > 0; i--) {
+        history[i] = history[i - 1];
+    }
```

Comment:
Probably could just change this to use std::vector's insert() and pop_back() function if you wanted to simplify this.